File: example/src/main/java/eu/darken/mvpbakery/example/screens/MainActivity.java
package eu.darken.mvpbakery.example.screens;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import java.util.Locale;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import eu.darken.mvpbakery.base.MVPBakery;
import eu.darken.mvpbakery.base.StateForwarder;
import eu.darken.mvpbakery.base.ViewModelRetainer;
import eu.darken.mvpbakery.example.R;
import eu.darken.mvpbakery.injection.ComponentSource;
import eu.darken.mvpbakery.injection.InjectedPresenter;
import eu.darken.mvpbakery.injection.ManualInjector;
import eu.darken.mvpbakery.injection.PresenterInjectionCallback;
import eu.darken.mvpbakery.injection.fragment.HasManualFragmentInjector;


public class MainActivity extends AppCompatActivity implements MainPresenter.View, HasManualFragmentInjector {

    @Inject ComponentSource<Fragment> componentSource;
    @Inject MainPresenter presenter;
    @BindView(R.id.container) ViewPager viewPager;
    @BindView(R.id.bindcounter) TextView bindCounter;
    @BindView(R.id.pageinfo) TextView pageInfo;

    private final StateForwarder stateForwarder = new StateForwarder();
    private PagerAdapter pagerAdapter;

    @Override
    public ManualInjector<Fragment> supportFragmentInjector() {
        return componentSource;
    }

    @Override
    protected void onCreate(Bundle si) {
        super.onCreate(si);
        stateForwarder.onCreate(si);
        MVPBakery.<MainPresenter.View, MainPresenter>builder()
                .stateForwarder(stateForwarder)
                .presenterFactory(new InjectedPresenter<>(this))
                .presenterRetainer(new ViewModelRetainer<>(this))
                .addPresenterCallback(new PresenterInjectionCallback<>(this))
                .attach(this);

        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        pagerAdapter = new PagerAdapter(getSupportFragmentManager(), si != null ? si.getInt("pages", 0) : 0);
        viewPager.setOffscreenPageLimit(pagerAdapter.getCount());
        viewPager.setAdapter(pagerAdapter);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                final Fragment fragment = pagerAdapter.getItem(position);
                pageInfo.setText(fragment.toString() + "\n");
                pageInfo.append(String.valueOf(position + 1) + "/" + pagerAdapter.getCount());

            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putInt("pages", pagerAdapter.getCount());
        stateForwarder.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void showFragment(Class<? extends Fragment> fragmentClass) {
        pagerAdapter.addPage(true);
        viewPager.setOffscreenPageLimit(pagerAdapter.getCount());
    }

    @Override
    public void showBinderCounter(int count) {
        bindCounter.setText(String.format(Locale.US, "Activity Rotation Count: %d", count));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add_fragment:
                presenter.addFragment();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: example/src/main/java/eu/darken/mvpbakery/example/screens/MainActivity.kt
package eu.darken.mvpbakery.example.screens

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.ButterKnife
import eu.darken.mvpbakery.base.MVPBakery
import eu.darken.mvpbakery.base.StateForwarder
import eu.darken.mvpbakery.base.ViewModelRetainer
import eu.darken.mvpbakery.example.R
import eu.darken.mvpbakery.injection.ComponentSource
import eu.darken.mvpbakery.injection.InjectedPresenter
import eu.darken.mvpbakery.injection.ManualInjector
import eu.darken.mvpbakery.injection.PresenterInjectionCallback
import eu.darken.mvpbakery.injection.fragment.HasManualFragmentInjector
import java.util.*
import javax.inject.Inject


class MainActivity : AppCompatActivity(), MainPresenter.View, HasManualFragmentInjector {

    @Inject
    lateinit var componentSource: ComponentSource<Fragment>
    @Inject
    lateinit var presenter: MainPresenter
    @BindView(R.id.container)
    lateinit var viewPager: ViewPager
    @BindView(R.id.bindcounter)
    lateinit var bindCounter: TextView
    @BindView(R.id.pageinfo)
    lateinit var pageInfo: TextView

    private val stateForwarder = StateForwarder()
    private var pagerAdapter: PagerAdapter? = null

    override fun supportFragmentInjector(): ManualInjector<Fragment> = componentSource

    override fun onCreate(si: Bundle?) {
        super.onCreate(si)
        stateForwarder.onCreate(si)
        MVPBakery.builder<MainPresenter.View, MainPresenter>()
                .stateForwarder(stateForwarder)
                .presenterFactory(InjectedPresenter(this))
                .presenterRetainer(ViewModelRetainer(this))
                .addPresenterCallback(PresenterInjectionCallback(this))
                .attach(this)

        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        pagerAdapter = PagerAdapter(this, supportFragmentManager)
        viewPager.offscreenPageLimit = pagerAdapter!!.count
        viewPager.adapter = pagerAdapter
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) = Unit

            override fun onPageSelected(position: Int) {
                val fragment = pagerAdapter!!.getItem(position)
                pageInfo.text = fragment.toString() + "\n"
                pageInfo.append((position + 1).toString() + "/" + pagerAdapter!!.count)
            }

            override fun onPageScrollStateChanged(state: Int) = Unit
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt("pages", pagerAdapter!!.count)
        stateForwarder.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun addFragment(fragmentClass: Class<out Fragment>) {
        pagerAdapter!!.addPage(fragmentClass, true)
        viewPager.offscreenPageLimit = pagerAdapter!!.count
        viewPager.currentItem = pagerAdapter!!.count - 1
    }

    override fun showBinderCounter(count: Int) {
        bindCounter.text = String.format(Locale.US, "Activity Rotation Count: %d", count)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.add_fragment -> {
                presenter.addNewFragment()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
